### PR TITLE
Fix truncating of results when sorted by relational field

### LINKED ISSUE #22785: Query limit with a relational sort will not return the correct amount
### Describe the Bug

When you have a collection with a o2m relational sort this would return less than what in the collection. When I would query the collection and sort it based on a field inside of a relation. With a limit `?limit=1000` that exceeds the total items in my collection it will not return the total amount of the collection.

### To Reproduce

When you have a table with a relational sort this would return less than what in the collection. As an example you have a collection and a relation with languages. You have multiple translations for said collection. When you would query this collection to sort on every translated name. 

If you have 500 records and you set a limit `?limit=1000&sort=collection.language_translation&fields=code`
This will not return 500 records.

I have added a test db here:
[relational-sort-with-limit-not-working.sql.zip](https://github.com/user-attachments/files/15891133/relational-sort-with-limit-not-working.sql.zip)

You can test this with setting the limit to a number that exceeds the current number. As example:
http://localhost:8055/items/country?limit=6&fields=code&fields=name&sort=translations.country_name_translation&fields=code

As there are 5 items and this limits 6 it should return 5 but it returns 3.
With the fix it returns all 5

### Directus Version

v10.11.2

### Hosting Strategy

Self-Hosted (Docker Image)

### Database

PostgreSQL 13

### PR DESCRIPTION
**Pull request by VincentKempers**
## Scope

What's changed:

- ~Adds the environment variable `QUERY_LIMIT_MAX` to the inner sort query~
  - ~when this max has been set it will use this for the inner query~
  - ~when it's not set or -1 it will clear the limit which would allow the sort to return the accurate amount of records on sorts that has a relational field.~
- If we are doing a multi relational sort, instead of sorting the results by the sort fields again in the inner query, we rely on the row_number for order. This allows us to keep the limit in the inner query in place and should be faster than removing the limit overall

### Explanation

The `directus_row_number` is present to facilitate relational sorting. A row with `directus_row_number = 1` is the correctly sorted first result for each primary key. That is why there is the `where inner.directus_row_number = 1` clause added to the wrapper query as that is the only result we are interested in when compiling all sorted results.

Instead of ordering the inner query by the sort fields again, we know that each partition is already correctly sorted within itself, so it can just be sorted by the row number in an ascending order, as that is guaranteed to return the top result per primary key. Then if a limit is applied it only cuts off the results that actually are more than the query limit, and not some other results.

The overall order of all results is still maintained by the outer order by clause of the wrapper query, that sorts by the sort value passed out from the inner query.

### Examples using the provided example dataset

**Results of the inner query, before applying this fix**

Query: 

```SQL
select distinct "country"."code",
                "tfnix"."country_name_translation"                                                                as "sort_jofrt",
                row_number()
                over (partition by "country"."code" order by "tfnix"."country_name_translation" asc)              as "directus_row_number"
from "country"
         left join "country_translations" as "tfnix" on "country"."code" = "tfnix"."country_code"
order by "tfnix"."country_name_translation" asc
limit 6
```

|code|sort_jorft |directus_row_number|
|----|-----------|-------------------|
|US  |Amerika    |1                  |
|US  |Amerika    |2                  |
|DE  |Deutschland|1                  |
|DE  |Duitsland  |2                  |
|DE  |Germany    |3                  |
|JM  |Jamaica    |1                  |

You can see that the row number is all over the place, and only 3 results with the row number `1` is actually returned by the limited query.

**Results of the inner query, after applying this fix**

Query: 

```SQL 
select distinct "country"."code",
                "tfnix"."country_name_translation"                                                                as "sort_jofrt",
                row_number()
                over (partition by "country"."code" order by "tfnix"."country_name_translation" asc)              as "directus_row_number"
from "country"
         left join "country_translations" as "tfnix" on "country"."code" = "tfnix"."country_code"
order by "directus_row_number" asc, "tfnix"."country_name_translation" asc
limit 6
```

| code | sort\_jofrt | directus\_row\_number |
| :--- | :--- | :--- |
| DE | Deutschland | 1 |
| JM | Jamaica | 1 |
| NL | Nederland | 1 |
| SR | Surinam | 1 |
| US | Amerika | 1 |
| DE | Duitsland | 2 |

## Potential Risks / Drawbacks

- Quite sure this will work on different databases, but so far only tested in Postgres and SQLite

## Review Notes / Questions

- I have added a test db here:
[relational-sort-with-limit-not-working.sql.zip](https://github.com/user-attachments/files/15890550/relational-sort-with-limit-not-working.sql.zip)

You can test this with setting the limit to a number that exceeds the current number. As example:
http://localhost:8055/items/country?limit=6&fields=code&fields=name&sort=translations.country_name_translation&fields=code

as there are 5 items and this limits 6 it should return 5 but it returns 3.
With the fix it returns all 5

---

Fixes #22785

### CODE REVIEW

**Review by joselcvarela — DISMISSED** [review]
This does not look the best solution as if the inner query table has millions of records, the query can take some time to compute.
Although, I don't see other solutions.
But, at least, it's using the `QUERY_LIMIT_MAX` so if there's any issues, users can adjust this configuration.

**Comment by hanneskuettner** [comment]
After talking to @VincentKempers I updated this PR to use a different approach. I realized that the inner query is already ordered within each partition because of the `directus_row_number` so we can order by that in the inner query, apply the where clause in the outer query like before and we are good to go, afaict.

File: api/src/database/run-ast.ts [hunk 1]
```diff
@@ -334,7 +332,16 @@ async function getDBQuery(
 					orderByString,
 					orderByFields,
 				);
+
+				// Only order by directus_row_number, as the directus_row_number is derived from a window function that is
+				// ordered by the sort fields within every primary key partition. Since the only relevant result is the first
+				// row of this partition, the order by here ensures that all rows with a directus_row_number = 1 show up first
+				// in the inner query result, and are correctly truncated by the limit, but not earlier.
+				orderByString = `?? asc, ${orderByString}`;
+				orderByFields.unshift(knex.ref('directus_row_number'));
```
Comment (licitdev) [inline api/src/database/run-ast.ts:341]: @hanneskuettner Love this fantastic solution to sort by the row number. I've patched it by including the original query sorting. For your review please~ 🫡

Comment (hanneskuettner) [reply api/src/database/run-ast.ts:343]: Very good find! As discussed privately, adding the original query sorting, in addition to the row number, will ensure that the the results are correctly sorted in the inner query, if there are ever more results than the query limit, which was overlooked by me 👍

File: api/src/database/run-ast.ts [hunk 2]
```diff
@@ -359,7 +368,7 @@ async function getDBQuery(
 		.from(table)
 		.innerJoin(knex.raw('??', dbQuery.as('inner')), `${table}.${primaryKey}`, `inner.${primaryKey}`);
 
-	if (sortRecords && needsInnerQuery) {
+	if (sortRecords) {
```
[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/database/run-ast.ts:371]: Yes, it's definitely safe to remove! 👍👍

**Review by hanneskuettner — APPROVED** [review]

**Review by licitdev — APPROVED** [review]

**Pull request merged**